rpcs3/Emu/Cell/SPULLVMRecompiler.cpp

```diff
@@ -5297,36 +5316,29 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
 
 	void FREST(spu_opcode_t op)
 	{
-		// TODO
-		if (g_cfg.core.spu_xfloat_accuracy == xfloat_accuracy::accurate)
+		register_intrinsic("spu_frest", [&](llvm::CallInst* ci)
 		{
-			const auto a = get_vr<f32[4]>(op.ra);
-			const auto mask_ov = sext<s32[4]>(bitcast<s32[4]>(fabs(a)) > splat<s32[4]>(0x7e7fffff));
-			const auto mask_de = eval(noncast<u32[4]>(sext<s32[4]>(fcmp_ord(a == fsplat<f32[4]>(0.)))) >> 1);
-			set_vr(op.rt, (bitcast<s32[4]>(fsplat<f32[4]>(1.0) / a) & ~mask_ov) | noncast<s32[4]>(mask_de));
-			return;
-		}
+			const auto a = bitcast<u32[4]>(value<f32[4]>(ci->getOperand(0)));
 
-		if (g_cfg.core.spu_xfloat_accuracy == xfloat_accuracy::approximate)
-		{
-			register_intrinsic("spu_frest", [&](llvm::CallInst* ci)
-			{
-				const auto a = value<f32[4]>(ci->getOperand(0));
-				// Gives accuracy penalty, frest result is within one newton-raphson iteration for accuracy
-				const auto approx_result = fsplat<f32[4]>(0.999875069f) / a;
-				// Zeroes the last 11 bytes of the mantissa so FI calculations end up correct if needed
-				return bitcast<f32[4]>(bitcast<u32[4]>(approx_result) & splat<u32[4]>(0xFFFFF800));
-			});
-		}
-		else
-		{
-			register_intrinsic("spu_frest", [&](llvm::CallInst* ci)
+			const auto a_fraction = (a >> splat<u32[4]>(18)) & splat<u32[4]>(0x1F);
+			const auto a_exponent = (a >> splat<u32[4]>(23)) & splat<u32[4]>(0xFF);
+			const auto a_sign = (a & splat<u32[4]>(0x80000000));
+			value_t<u32[4]> final_result = eval(splat<u32[4]>(0));
+
+			for (u32 i = 0; i < 4; i++)
 			{
-				const auto a = value<f32[4]>(ci->getOperand(0));
-				// Fast but this makes the result vary per cpu
-				return fre(a);
-			});
-		}
+				const auto eval_fraction = eval(extract(a_fraction, i));
+				const auto eval_exponent = eval(extract(a_exponent, i));
+				const auto eval_sign = eval(extract(a_sign, i));
+
+				value_t<u32> r_fraction = load_const<u32>(m_spu_frest_fraction_lut, eval_fraction);
+				value_t<u32> r_exponent = load_const<u32>(m_spu_frest_exponent_lut, eval_exponent);
```

Comment:
what happened to the gather version?

Reply:
I looked into a bit more, the only intrinsic that'd make sense for use is llvm.masked.gather (Intrinsic::masked_gather) and it pretty much what I do except it requires a vector of pointers as a starting point(see https://llvm.org/docs/LangRef.html#llvm-masked-gather-intrinsics ). Of note is that it describes what the Instrinsic with all true mask results in which is very similar to what I already do(except I do the pointer calculations in the middle).

Reply:
So what's preventing this solution from being used here? You can load the base pointers into a 512 bit vector, add index, and load.

Reply:
If the amount of pointers (8) is the problem you can try to force the upper 4 indices to be equal to the lowest index so only 4 memory locations are accessed.
It should be tested if it benefits performance.

Reply:
I tried Intrinsic::vp_gather which seemed as good(though it returns an array of u64) but I can't seem to get it to work, it crashes during IR production from parsing the parameters than I'm giving it. If someone manages to make it work they're welcome to try.